enhancements/kubelet/virtualization-higher-workload-density.md

```diff
@@ -0,0 +1,502 @@
+| Phase 1 - Out-Of-Tree SWAP with WASP TP                            | mid 2024     | Tech-Preview      | Beta                 | Manual                                 |
+| Phase 2 - WASP GA with swap-based evictions                        | end 2024     | GA                | Beta                 | Manual                                 |
+| Phase 3 - Transition to Kubernetes SWAP. Limited to CNV-only users | mid-end 2025 | Start Deprecation | GA                   | Manual                                 |
+| Phase 4 - Kubernetes SWAP for all OpenShift users                  | TBD          | Removed           | GA                   | Operator                               |
+
+Because [Kubernetes SWAP] is currently in Beta and is only expected to GA within
+Kubernetes releases 1.33-1.35 (discussion about its GA criterias is still ongoing).
+this proposal is taking a four-phased approach in order to meet the timeline requirements.
+
+* **Phase 1** - OpenShift Virtualization will provide an out-of-tree
+  solution (WASP) to enable higher workload density and swap-based eviction mechanisms.
+* **Phase 2** - WASP will include swap-based eviction mechanisms.
+* **Phase 3** - OpenShift Virtualization will transition to [Kubernetes SWAP] (in-tree).
+  OpenShift will [allow](#swap-ga-for-cnv-users-only) SWAP to be configured only if OpenShift Virtualization is installed on the cluster.
+  In this phase, WASP deprecation will start.
+* **Phase 4** - OpenShift will GA SWAP for every user, even if OpenShift Virtualization
+  is not installed on the cluster. In this phase WASP will be removed, machine-level configuration
+  will be managed by Machine Config Operator.
+
+### Workflow Description
+
+**cluster administrator** (Cluster Admin) is a human user responsible
+for deploying and administering a cluster.
+
+**virtual machine owner** (VM Owner) is a human user operating a
+virtual machine in a cluster.
+
+#### Workflow: Configuring higher workload density
+
+1. The cluster admin is deploying a bare-metal OpenShift cluster
+2. The cluster admin is deploying the OpenShift Virtualization Operator
+3. The cluster admin is deploying the WASP Agent according to the
+   [documentation](https://github.com/openshift-virtualization/wasp-agent/blob/main/docs/configuration.md)
+4. The cluster admin is configuring OpenShift Virtualization for higher
+   workload density via:
+
+   a. the OpenShift Virtualization Console "Settings" page 
+   b. [or `HCO` API](https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/docs/cluster-configuration.md#configure-higher-workload-density)
+
+The cluster is now set up for higher workload density.
+
+In phase 3, deploying the WASP agent will not be needed.
+
+#### Workflow: Leveraging higher workload density
+
+1. The VM Owner is creating a regular virtual machine and is launching it. The VM owner must not specify memory requests in the VM spec, but only the guest memory size.
+
+### API Extensions
+
+#### Phase 3
```

Comment:
this phase glosses over some other MC pieces that are needed (to actually provision swap). that's part of the motivation of doing SwapOperator: to couple the actual swap provisioning with pointing the kubelet to use swap

Reply:
@haircommander Good point! Let me then suggest the following - since in phase 3 kube swap will be GAed, but wasp-agent will still be there (deprecated but not removed), the user actually doesn't need to move to kube swap right away. This transition can happen in phase 4 by the swap operator (and its even better, less error prone). 
Therefore transition to phase 3 can be no-op from user perspective.